test/e2e/storage/host_path_type.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = utils.SIGDescribe("HostPathType", func() {
+	f := framework.NewDefaultFramework("host-path-type")
+
+	var (
+		ns           string
+		hostBaseDir  string
+		mountBaseDir string
+		basePod      *v1.Pod
+
+		hostPathUnset             v1.HostPathType = v1.HostPathUnset
+		hostPathDirectoryOrCreate v1.HostPathType = v1.HostPathDirectoryOrCreate
+		hostPathDirectory         v1.HostPathType = v1.HostPathDirectory
+		hostPathFileOrCreate      v1.HostPathType = v1.HostPathFileOrCreate
+		hostPathFile              v1.HostPathType = v1.HostPathFile
+		hostPathSocket            v1.HostPathType = v1.HostPathSocket
+		hostPathCharDev           v1.HostPathType = v1.HostPathCharDev
+		hostPathBlockDev          v1.HostPathType = v1.HostPathBlockDev
+	)
+
+	BeforeEach(func() {
+		ns = f.Namespace.Name
+
+		By("Create a pod for further testing")
+		hostDir := "/tmp/hostpath-test"
+		mountDir := "/mnt/test"
+		basePod = f.PodClient().CreateSync(newHostPathTypeTestPod("", hostDir, mountDir, &hostPathDirectoryOrCreate))
+
+		By("Create a temporary clean directory for further testing")
+		result, err := utils.PodExec(basePod, fmt.Sprintf("mktemp -d -p %s hostpath.XXXXXX", mountDir))
+		Expect(err).NotTo(HaveOccurred())
+
+		mountBaseDir = strings.TrimSpace(result)
+		hostBaseDir = path.Join(hostDir, strings.TrimPrefix(mountBaseDir, mountDir))
+	})
+
+	Context("HostPathType Directory Test", func() {
+		var targetDir string
+
+		BeforeEach(func() {
+			targetDir = path.Join(hostBaseDir, "adir")
+
+			By("Should fail on mounting non-existent directory 'adir' when HostPathType is HostPathDirectory")
```

Comment:
Can each HostPathType be a separate It() instead of inside the BeforeEach()?

Reply:
Also should we have test cases where the hostpath is of the wrong type?  For example, hostpath is a file but type is HostPathDirectory.

Reply:
We do have separate `It()`. Here `BeforeEach()` is mainly used to prepare the needed environment.

Reply:
Yeah, wrong hostpath type is not well tested. Will cover that.

Reply:
I guess it's a little strange the the By() in the BeforeEach() could actually be a standalone test case too.

Reply:
I think this particular scenario of failing when the directory doesn't exist should be its own test case (a separate `It()`) instead of in the `BeforeEach()`.  You could use a different directory like "does-not-exist-dir".  Same for all the other types